### PR TITLE
VPN-6406: Improve iOS connection health

### PR DESCRIPTION
**Pull request by mcleinman**
## Description

- Adds new state of 'pending' when the daemon starts, as well as when there is no internet access
- Added check for internet access
- Added metrics around new pending status, as well as to check silent server switches (to be used in the rest of this epic)

## Reference

VPN-6406

## Checklist
    
- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [x] I have added thorough tests where needed

### CODE REVIEW

**Comment by mcleinman** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

These are a handful of metrics to learn more about how people use the VPN app. How often do they run into internet connection issues? How often are there errors when sending pings to check connection health? How often does the daemon try to switch the server (to resolve connection issues)?

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? 

We want to strengthen code around switching servers, but need some basic data around how often the users are running into issues, and what types.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

Carefully considered which metrics were needed. Considered just making the changes to code without data, but that comes with risk of significant battery usage, and thus we wanted to get some data.

4) Can current instrumentation answer these questions?

No, unfortunately.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Name</td>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>changed_to_pending</td>
    <td>Event that is recorded when connection status changes to pending</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-6406">VPN-6406</a></td>
  </tr>
  <tr>
    <td>pending_count</td>
    <td>Count of health checks which result in status is pending (bad internet)</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-6406">VPN-6406</a></td>
  </tr>
  <tr>
    <td>pending_time</td>
    <td>Time distribution of how long connectivity status stays in pending (bad internet)</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-6406">VPN-6406</a></td>
  </tr>
  <tr>
    <td>ping_analyzer_error</td>
    <td>Event for how often ping analyzer returns an error (expect this to be very, very rare)</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-6406">VPN-6406</a></td>
  </tr>
  <tr>
    <td>daemon_silent_server_switch</td>
    <td>Event for daemon-led silent server switches</td>
    <td>technical</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-6406">VPN-6406</a></td>
  </tr>

</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

This collection is documented in the Glean Dictionary at https://dictionary.telemetry.mozilla.org/"

7) How long will this data be collected?  Choose one of the following:

Several months, with potential of requesting renewal down the line. mcleinman@mozilla.com

8) What populations will you measure?

No filters - all channels, countries, and locales, unless the user has opted out of data collection on that device.

9) If this data collection is default on, what is the opt-out mechanism for users?

When launching the app for the first time, a user is given an option for whether to allow data collection. After this initial set up screen, a user can always toggle data collection permissions in the System Preferences screen.

10) Please provide a general description of how you will analyze this data.

Dashboards that engineers and others will consult on a regular basis.

11) Where do you intend to share the results of your analysis?

Within the Mozilla VPN team and PXI Products team.

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? 

No

File: src/platforms/ios/InternetChecker.swift
Comment (oskirby) [inline src/platforms/ios/InternetChecker.swift:11]: I am a little concerned that this is a little too aggressive and inflexible. IIUC, this is sending a ping to the wireguard server's public IPv4 address to verify internet connectivity and 100ms is not going to be enough to reach a server on the other side of the internet.

Comment (mcleinman) [reply src/platforms/ios/InternetChecker.swift:11]: Discussed in a meeting this morning; I'll add some more context in code comments..

**Comment by mcleinman** [comment]
Changes have been made per our discussion, this is ready for review. (I'm leaving it on draft until we get data steward approval, but code review can happen spearately.)

**Comment by mcleinman** [comment]
Adding @adudenamedruby for data review, as I believe Travis is heading to a work week. Feel free to unassign yourself if you don't have capacity, and can I request assistance from another steward. Thanks!

**Comment by adudenamedruby** [comment]
r+
# Data Review Form

_1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?_ 

Yes, telemetry is documented at https://dictionary.telemetry.mozilla.org/apps/mozilla_vpn

_2) Is there a control mechanism that allows the user to turn the data collection on and off?_ 

Yes.

_3) If the request is for permanent data collection, is there someone who will monitor the data over time?_

This will be reviewed by a data-steward every 6 months.

_4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?_

Type 1

_5) Is the data collection request for default-on or default-off?_

Default on; user is asked on first app launch.

_6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?_

No

_7) Is the data collection covered by the existing Firefox privacy notice?_ 

Yes

_8) Does the data collection use a third-party collection tool?_ 

No

**Review by oskirby — APPROVED** [review]
I think this looks good.

**Pull request merged**
